### PR TITLE
GH-43554: [Go] Handle excluded fields

### PR DESCRIPTION
**Pull request by tscottcoombes1**
### Rationale for this change

We want to be able to handle excluded fields.

### What changes are included in this PR?

* we no longer use the value of the field when getting the element type of a list (as the values are invalid for excluded fields)
* similarly for map, key value pairs, we don't use the value is there is none
* add some tests 

### Are these changes tested?

yes

### Are there any user-facing changes?

no

* GitHub Issue: #43554

### CODE REVIEW

File: go/arrow/util/protobuf_reflect.go
Comment (tscottcoombes1) [inline go/arrow/util/protobuf_reflect.go:240]: this is the key change for lists, protoreflect doesn't come with a get underlying type function instead the isList is set to true, so on the 2nd pass to get the data type we override the isList with isListItem

File: go/arrow/util/protobuf_reflect_test.go
Comment (joellubi) [inline go/arrow/util/protobuf_reflect_test.go:493]: Any reason not to `CheckRecord()` here too?

File: go/arrow/util/protobuf_reflect.go
Comment (joellubi) [inline go/arrow/util/protobuf_reflect.go:414]: May be a bit nicer instead to `return` early from the conditional and get rid of the need for `else`, since this block already has considerable nesting.

```suggestion
			return
		}
		for _, k := range pmr.rValue.MapKeys() {
```

**Review by joellubi — COMMENTED** [review]
Looks good, just a few comments

File: go/arrow/util/protobuf_reflect_test.go
Comment (tscottcoombes1) [reply go/arrow/util/protobuf_reflect_test.go:483]: nope, I have added them

**Review by joellubi — APPROVED** [review]
Thanks for contributing!

**Comment by tscottcoombes1** [comment]
Hi @zeroshade can I get your code owner review as well please :)

File: go/arrow/util/protobuf_reflect_test.go
Comment (zeroshade) [inline go/arrow/util/protobuf_reflect_test.go:239]: Does this comment need to be here? Can we remove it?

**Review by zeroshade — APPROVED** [review]
Just a single nitpick, after we address it then we can merge this

File: go/arrow/util/protobuf_reflect_test.go
Comment (tscottcoombes1) [reply go/arrow/util/protobuf_reflect_test.go]: removed

**Pull request merged**
